File: app/src/main/java/com/muratcan/apps/petvaccinetracker/repository/PetRepository.java
package com.muratcan.apps.petvaccinetracker.repository;

import android.app.Application;

import androidx.lifecycle.LiveData;

import com.muratcan.apps.petvaccinetracker.database.AppDatabase;
import com.muratcan.apps.petvaccinetracker.model.Pet;
import com.muratcan.apps.petvaccinetracker.model.Vaccine;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class PetRepository {
    private final AppDatabase database;
    private final ExecutorService executorService;

    public interface OnSuccessListener {
        void onSuccess();
    }

    public interface OnErrorListener {
        void onError(String error);
    }

    public interface OnVaccinesLoadedListener {
        void onVaccinesLoaded(List<Vaccine> vaccines);
    }

    public interface OnPetLoadedListener {
        void onPetLoaded(Pet pet);
    }

    public PetRepository(Application application) {
        database = AppDatabase.getInstance(application);
        executorService = Executors.newSingleThreadExecutor();
    }

    public LiveData<List<Vaccine>> getVaccinesForPetLive(long petId) {
        return database.vaccineDao().getVaccinesForPet(petId);
    }

    public void getVaccinesForPet(long petId, OnVaccinesLoadedListener successListener, OnErrorListener errorListener) {
        executorService.execute(() -> {
            try {
                List<Vaccine> vaccines = database.vaccineDao().getVaccinesForPetSync(petId);
                successListener.onVaccinesLoaded(vaccines);
            } catch (Exception e) {
                errorListener.onError("Error loading vaccines: " + e.getMessage());
            }
        });
    }

    public LiveData<List<Pet>> getAllPets() {
        return database.petDao().getAllPets();
    }

    public void addPet(Pet pet, OnSuccessListener successListener, OnErrorListener errorListener) {
        executorService.execute(() -> {
            try {
                long id = database.petDao().insert(pet);
                pet.setId(id);
                successListener.onSuccess();
            } catch (Exception e) {
                errorListener.onError("Error adding pet: " + e.getMessage());
            }
        });
    }

    public void getPetById(long petId, OnPetLoadedListener successListener, OnErrorListener errorListener) {
        executorService.execute(() -> {
            try {
                Pet pet = database.petDao().getPetById(petId);
                if (pet != null) {
                    successListener.onPetLoaded(pet);
                } else {
                    errorListener.onError("Pet not found");
                }
            } catch (Exception e) {
                errorListener.onError("Error loading pet: " + e.getMessage());
            }
        });
    }
} 

File: app/src/main/java/com/muratcan/apps/petvaccinetracker/repository/PetRepository.kt
package com.muratcan.apps.petvaccinetracker.repository

import android.content.Context
import com.muratcan.apps.petvaccinetracker.database.AppDatabase
import com.muratcan.apps.petvaccinetracker.model.Pet
import com.muratcan.apps.petvaccinetracker.model.Vaccine
import com.muratcan.apps.petvaccinetracker.model.VaccineHistoryItem
import com.muratcan.apps.petvaccinetracker.model.VaccineWithPetName
import com.muratcan.apps.petvaccinetracker.util.DefaultDispatcherProvider
import com.muratcan.apps.petvaccinetracker.util.DispatcherProvider
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext

class PetRepository(
    context: Context,
    private val dispatchers: DispatcherProvider = DefaultDispatcherProvider()
) {
    private val database = AppDatabase.getDatabase(context)
    private val petDao = database.petDao()
    private val vaccineDao = database.vaccineDao()

    fun getAllPetsFlow(): Flow<List<Pet>> = petDao.getAllPetsFlow()

    fun getPetsForUser(userId: String): Flow<List<Pet>> = petDao.getPetsForUser(userId)

    suspend fun getPetByIdFlow(petId: Long): Pet? = withContext(dispatchers.io) {
        petDao.getPetById(petId)
    }

    fun getVaccinesForPetFlow(petId: Long): Flow<List<Vaccine>> =
        vaccineDao.getVaccinesForPetFlow(petId)

    suspend fun addPet(pet: Pet): Long = withContext(dispatchers.io) {
        petDao.insert(pet)
    }

    suspend fun updatePet(pet: Pet) = withContext(dispatchers.io) {
        petDao.update(pet)
    }

    suspend fun deletePet(pet: Pet) = withContext(dispatchers.io) {
        petDao.delete(pet)
    }

    suspend fun addVaccine(vaccine: Vaccine): Long = withContext(dispatchers.io) {
        vaccineDao.insert(vaccine)
    }

    suspend fun updateVaccine(vaccine: Vaccine) = withContext(dispatchers.io) {
        vaccineDao.update(vaccine)
    }

    suspend fun deleteVaccine(vaccine: Vaccine) = withContext(dispatchers.io) {
        vaccineDao.delete(vaccine)
    }

    fun getVaccineHistory(): Flow<List<VaccineHistoryItem>> {
        return vaccineDao.getVaccineHistoryRaw()
            .map { vaccineWithPetList: List<VaccineWithPetName> ->
                vaccineWithPetList.map { vaccineWithPet: VaccineWithPetName ->
                    vaccineWithPet.toHistoryItem()
                }
            }
    }

    fun getVaccineHistoryForPet(petId: Long): Flow<List<VaccineHistoryItem>> {
        return vaccineDao.getVaccineHistoryForPetRaw(petId)
            .map { vaccineWithPetList: List<VaccineWithPetName> ->
                vaccineWithPetList.map { vaccineWithPet: VaccineWithPetName ->
                    vaccineWithPet.toHistoryItem()
                }
            }
    }
}